src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/TextBasedPhysicalPlanPrinterImpl2.java

```diff
@@ -0,0 +1,334 @@
+		out.append( indentLevelString + "bindJoin (" + op.getID() + ") " );
+		final LogicalOperator lop = op.getLogicalOperator();
+		printLogicalOperator( nameOfLogicalOp(lop), lop.getID(), out, indentLevelStringForOpDetail );
+		if ( lop instanceof LogicalOpTPAdd ) {
+			printFederationMember( ((LogicalOpTPAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		else if ( lop instanceof LogicalOpTPOptAdd ) {
+			printFederationMember( ((LogicalOpTPOptAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+
+		}
+		printSPARQLGraphPattern( ((LogicalOpTPAdd) lop).getTP(), indentLevelStringForOpDetail + singleBase, out );
+	}
+	
+	protected void printOperatorInfoForBindJoinWithFILTER ( final PhysicalOpBindJoinWithFILTER op, final PrintStream out, final String indentLevelString, final String indentLevelStringForOpDetail ) {
+		out.append( indentLevelString + "FILTERBindJoin (" + op.getID() + ") " );
+		final LogicalOperator lop = op.getLogicalOperator();
+		printLogicalOperator( nameOfLogicalOp(op.getLogicalOperator()), op.getLogicalOperator().getID(), out, indentLevelStringForOpDetail );
+		if ( lop instanceof LogicalOpTPAdd ) {
+			printFederationMember( ((LogicalOpTPAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+			printSPARQLGraphPattern( ((LogicalOpTPAdd) lop).getTP(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		else if ( lop instanceof LogicalOpTPOptAdd ) {
+			printFederationMember( ((LogicalOpTPOptAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+			printSPARQLGraphPattern( ((LogicalOpTPAdd) lop).getTP(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		else if (lop instanceof LogicalOpBGPAdd ) {
+			printFederationMember( ((LogicalOpBGPAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+			printSPARQLGraphPattern( ((LogicalOpBGPAdd) lop).getBGP(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		else if (lop instanceof LogicalOpBGPOptAdd ) {
+			printFederationMember( ((LogicalOpBGPOptAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+			printSPARQLGraphPattern( ((LogicalOpBGPOptAdd) lop).getBGP(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		else if (lop instanceof LogicalOpGPAdd ) {
+			printFederationMember( ((LogicalOpGPAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+			printSPARQLGraphPattern( ((LogicalOpGPAdd) lop).getPattern(), indentLevelStringForOpDetail + singleBase, out );
+		}
+		else if (lop instanceof LogicalOpGPOptAdd ) {
+			printFederationMember( ((LogicalOpGPOptAdd) lop).getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+			printSPARQLGraphPattern( ((LogicalOpGPOptAdd) lop).getPattern(), indentLevelStringForOpDetail + singleBase, out );
+		}
```

Comment:
Move this code into a separate helper function that you can call from each of the `printOperatorInfoFor...` functions in which you currently have a copy of this code block.

One more thing that I have come across recently that makes writing such code with `instanceof` checks followed by a casting operation much easier and more readable: Since Java 14 it is possible to shorten, say, the last else-if block here as follows.
```java
else if ( lop instanceof LogicalOpGPOptAdd addOp ) {
	printFederationMember( addOp.getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
	printSPARQLGraphPattern( addOp.getPattern(), indentLevelStringForOpDetail + singleBase, out );
}
```
Notice the new variable called `addOp` introduced right after the `instanceof` test. This variable is assigned with the same Java object as `lop`, but automatically casted to `LogicalOpGPOptAdd`. As a consequence, the explicit casting inside the else-if block can be removed and the code is less complex.

Look at https://www.baeldung.com/java-pattern-matching-instanceof if you want to read a bit more about this new Java feature.

Reply:
Huanyu, I have fixed the issue. That is, I have increased the Java version to be used, both within our Maven POM file and our configuration file for the CI workflow. This means that you may have to refresh the project in your IDE. For instance, in Eclipse I had to right-click on the HeFQUIN project in the "Package Explorer" tab and the select "Refresh" in the pop-up menu.
/cc @keski